### DISCUSSION TITLE
Retrieve current user id | Extending with JSVM Hooks

### DISCUSSION THREAD

**Discussion by pauserratgutierrez** [pocketbase/pocketbase#6099] [Q&A] [answered]
Hello, I'm using Pocketbase 0.23.6 and extending with pb_hooks using JSVM. I need to get the current user id of the user triggering the record creation.

Basically, when the collection "locations" gets a create request, it adds the data. After the data is added successfuly I want to add more data to a "user_locations" collection.

```
// This works
onRecordCreateRequest(async (e) => {  
  try {
    const fetchGeolocation = require(`${__hooks}/utils/nominatim.js`);

    const locationText = e.record.get('location_text');
    if (!locationText) return;
    
    const { lat, lon, details } = await fetchGeolocation(locationText);

    if (!lat || !lon || !details) {
      console.error('Incomplete geolocation data received');
      return;
    }

    e.record.set('lat', lat);
    e.record.set('lon', lon);
    e.record.set('name', details.name);
    e.record.set('label', details.label);
    e.record.set('city', details.city);
    e.record.set('county', details.county);
    e.record.set('state', details.state);
    e.record.set('country', details.country);
    e.record.set('country_code', details.country_code);
    e.record.set('postcode', details.postcode);

    e.next();
  } catch (err) {
    console.error('Error fetching geolocation:', err);
    throw err;
  }
}, 'locations');

// Here I'm not able to get the user id
onRecordAfterCreateSuccess(async (e) => {
  try {
    const userID = e.auth()?.id;
    const locationID = e.record.id;

    let user_locations_collection = await $app.findCollectionByNameOrId('user_locations');
    if (!user_locations_collection) return;

    let record = new Record(user_locations_collection);

    record.set('user_id', userID);
    record.set('location_id', locationID);

    $app.save(record);

    e.next();
  } catch (err) {
    console.error('Error saving user location:', err);
    throw err;
  }
}, 'locations');
```

```
2024/12/12 20:31:15 Error saving user location: TypeError: Object has no member 'auth'
```

As context, I want to add that I'm doing it from the Pocketbase admin dashboard with my superuser administrator. Haven't tried with a normal auth user. But I would like to get the user id always if possible, independently of is done from the dashboard or the normal frontend.

Thank you in advance.

**Comment by ganigeorgiev** [comment, marked as answer]
_Edit: I see that you edited your post while I was writing mine, but the below is still valid._

There are several problems with _"I need to get the current user id"_:

1. `onRecordAfterCreateSuccess` is a record db hook, aka. it can be triggered by non-request actions, e.g. cli or any other programmatically invoked record save, and has no context about who or what has triggered it. `e.auth` is available ONLY as part of the `*Request` hooks.

2. It is not clear what is the type of the `user_id` field. If it is a `relation` then it can be associated only with a single collection, aka. you can't assign both superuser and regular user id because they are from 2 different auth collections.

Assuming that the above action should be ignored for superusers, if you want to retrieve the id of the authenticated user that submitted a create request, you can attach to the equivalent [request hook](https://pocketbase.io/docs/js-event-hooks/#request-hooks), for example:

```js
// e in the below hook "inherit" the request event
// see https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state
onRecordCreateRequest((e) => {
    e.next() // execute after the create response body has been committed

    if (!e.auth) {
        throw new UnauthorizedError()
    }

    if (e.hasSuperuserAuth()) {
        return // do nothing
    }

    try {
        // single retrieval function throw and none of the JSVM APIs return Promise so there is no need for async/await
        let user_locations_collection = e.app.findCollectionByNameOrId('user_locations');

        let record = new Record(user_locations_collection);

        record.set('user_id', e.auth.id);
        record.set('location_id', e.record.id);

        e.app.save(record);
    } catch (err) {
        e.app.logger().error('Error saving user location', "error", err);
        throw err; // do you need to rethrow?
    }
}, 'locations')
```

**Reply by pauserratgutierrez** [reply]
Thank you for the detailed explaination and for providing a working solution nearly instantly.
